#167917118 Add feature for user to see all trips and implement feedback

amkayondo opened the pull request:
#### What does this PR do?
Adds a feature for user to see all trips and implement feedback
**Feed Back Implemented**
- Modify error for a missing field
- Add No token error message
- user should get all trips with or without token
- Modify error code for accessing cancelled trip by user
- fix tests
#### Description of Task to be completed?
user can see all of his/her bookings
#### How should this be manually tested?
send `GET` - `/trips` 
#### Any background context you want to provide?
- `200` - OK
- `404` - NOT FOUND
- `403` - 
#### What are the relevant pivotal tracker stories?
[#167917118](https://www.pivotaltracker.com/story/show/167917118)
#### Screenshots 
![Screenshot (77)](https://user-images.githubusercontent.com/46356108/63115551-b0e0cc00-bf97-11e9-9eb7-abab8dd102c3.png)
![Screenshot (78)](https://user-images.githubusercontent.com/46356108/63115552-b0e0cc00-bf97-11e9-9703-2c1ac735075d.png)
![Screenshot (73)](https://user-images.githubusercontent.com/46356108/63115553-b1796280-bf97-11e9-8bc7-5f228a979e75.png)
![Screenshot (74)](https://user-images.githubusercontent.com/46356108/63115555-b1796280-bf97-11e9-9033-bc90f880efd6.png)
![Screenshot (75)](https://user-images.githubusercontent.com/46356108/63115556-b211f900-bf97-11e9-96ec-12615a0c70f5.png)
![Screenshot (76)](https://user-images.githubusercontent.com/46356108/63115559-b211f900-bf97-11e9-8c85-d14017bca4aa.png)

amkayondo committed:
ft(API): User can see their bookings
- Check if the user is admin to retrieve all bookings
- Get bookings made by only the user

amkayondo committed:
update heroku

amkayondo committed:
chore(API): Implement feedback
- Modify error for missing field
- Add No token error message
- user should get all trips with or without token
- Modify error code for accessing cancelled trip by user
- fix tests
[finishes #167917118]

Kasulejoseph commented on server/controllers/bookings/getAllBookings.js:
```diff
@@ -6,18 +6,18 @@ const book = new Book();
     const getUser = jwt.decode(req.headers.authorization);
-    if (getUser.isAdmin === true) {
-      const allBookings = await book.getAtllbookings;
+    if (getUser.isadmin === true) {
```
Why did you change this (`isAdmin`) from being a camelCase?
You can also better write the above with just `if (getUser.isadmin)`

Kasulejoseph commented on server/controllers/bookings/getAllBookings.js:
```diff
@@ -6,18 +6,18 @@ const book = new Book();
+    } resPonse.successData(res, 200, 'All bookings successfully fetched', forUser.rows);
   } catch (error) {
-    resPonse.errorMessage(res, 500, error.message);
+    console.log(error.message);
```
You're doing nothing to the error, remove the debugger statement and handle it correctly

Kasulejoseph commented on server/controllers/trips/createTrip.js:
```diff
@@ -15,7 +15,8 @@ const createTrip = async (req, res) => {
+      data.trip_date, data.status);
     const getTrip = trpExists.rows;
     if (getTrip.length === 0){
       const nwdata = await newtrip.creatAtrip(data);
```
What is `nwdata`? provide a better informative and descriptive variable name.

Kasulejoseph commented on server/controllers/trips/getTripbyId.js:
```diff
@@ -1,14 +1,18 @@
-    } resPonse.successData(res, 200, 'Trip found', foundTrip.rows[0]);
-  } catch (err) {}
+    if (admin.isadmin === true){
+      return resPonse.successData(res, 404, 'Trip found', foundTrip.rows[0]);
```
`Trip found` with a status code of `404`?

Kasulejoseph commented on server/controllers/trips/getTripbyId.js:
```diff
@@ -1,14 +1,18 @@
+      return resPonse.successData(res, 404, 'Trip found', foundTrip.rows[0]);
+    }
+  } catch (err) {
+    console.log(err.message);
```
Same here...

Kasulejoseph commented on server/controllers/trips/getTrips.js:
```diff
@@ -5,24 +5,17 @@ import Trip from '../../models/trips';
+    return resPonse.successDatas(res, 200, foundtrips.rows.length, foundtrips.rows);
+  }
+  // } catch (err) {
+  //   console.log(err.message);
```
Same here..

Kasulejoseph commented on server/middleware/filterTripId.js:
```diff
@@ -0,0 +1,22 @@
+  const notadmin = jwt.decode(req.headers.authorization);
+  const activeTrips = tripData.filter(x => x.status === 'cancelled');
+  if (!req.headers.authorization || req.headers.authorization === ''
+  || notadmin.isadmin === false || notadmin === null){
```
Could you write a reusable function which you just call here and other places that require authorization checks.

Kasulejoseph commented on server/middleware/filterTrips.js:
```diff
@@ -0,0 +1,21 @@
+  const notadmin = jwt.decode(req.headers.authorization);
+  const activeTrips = tripData.filter(x => x.status === 'active');
+  if (!req.headers.authorization || req.headers.authorization === ''
+  || notadmin.isadmin === false || notadmin === null){
```
That reusable function I talked about could just be called here as well. Try to make you're done as DRY as possible.

Kasulejoseph commented on server/middleware/validators/validateBookindId.js:
```diff
@@ -2,7 +2,7 @@ import resPonse from '../../helpers/responses/response';
 const Joi = require('@hapi/joi');
 
-const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].context.label}`));
+const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].message}`));
```
Similar blocks of code in 2 locations here and validateInput.js file, consider refactoring to make the code base DRY.

Kasulejoseph commented on server/middleware/validators/validateInput.js:
```diff
@@ -6,7 +6,7 @@ const Extension = require('@hapi/joi-date');
 const regeX2 = (/^[a-zA-Z0-9]{3,30}$/);
 const regexText = (/^[a-zA-Z]{3,30}$/);
-const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].context.label}`));
+const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].message}`));
```
Similar code as in validateBookindId.js file

Kasulejoseph commented on server/middleware/validators/validateBookindId.js:
```diff
@@ -2,7 +2,7 @@ import resPonse from '../../helpers/responses/response';
 
```
The filename has a typo `validateBookindId.js` fix it.

Kasulejoseph commented on server/middleware/validators/validateReq.js:
```diff
@@ -2,7 +2,7 @@ import resPonse from '../../helpers/responses/response';
 const Joi = require('@hapi/joi');
 
-const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].context.label}`));
+const validateError = (res, result) => resPonse.errorMessage(res, 400, (`${result.error.details[0].message}`));
```
Same code even here...

Kasulejoseph commented on server/tests/bookings.test.js:
```diff
@@ -46,7 +45,7 @@ before('NEW USER SIGN UP', (done) => {
       first_name: 'kayondo',
       last_name: 'edward',
```
Create a mock file for all these test data

Kasulejoseph commented on server/tests/bookings.test.js:
```diff
@@ -139,7 +138,7 @@ describe('BOOKINGS TESTS', () => {
       .set('Authorization', userToken)
       .end((err, res) => {
-        expect(res).to.have.status(200);
+        expect(res).to.have.status(404);
```
You need to assert for data in tests not only status codes..

Kasulejoseph commented on server/tests/bookings.test.js:
```diff
@@ -281,7 +277,7 @@ describe('BOOKINGS TESTS', () => {
       .set('Authorization', userToken)
       .end((err, res) => {
-        expect(res).to.have.status(200);
+        expect(res).to.have.status(404);
```
This looks like your just forcing tests to pass but in actual sense, they don't. why did it change from a 200 to a 404?

amkayondo committed:
chore(API) Implement feedback
- refactor code
- remove the debugger statements
- remove code duplication
[finishes #167917118]
